// UniSphere/src/pages/LandingPage.jsx
import { useState, useEffect } from "react";
import EventCard from "../components/EventCard";
import { fetchEvents } from "../api";

export default function LandingPage({ onRegister, eventRefreshKey }) {
  const [search, setSearch] = useState("");
  const [category, setCategory] = useState("All");
  const [sort, setSort] = useState("default");
  const [priceFilter, setPriceFilter] = useState("all"); // all, free, paid
  const [availabilityFilter, setAvailabilityFilter] = useState("all"); // all, available, full
  const [events, setEvents] = useState([]);
  const [countdown, setCountdown] = useState({ days: 0, hours: 0, minutes: 0, seconds: 0 });

  const defaultEvents = [
    {
      title: "Hackathon 2026",
      description: "Build amazing projects",
      category: "Technical",
      image: "https://www.cpduk.co.uk/sites/default/files/news-imported/cpd-product-marketing-alliance-best-practices-hackathon.jpg",
      date: "Feb 12 2026",
      time: "9:00 AM - 6:00 PM",
      venue: "Main Auditorium",
      coordinator: "Sanjay Karupothula",
      inspector: "Prof. Rajesh, IIT Hyderabad"
    },
    {
      title: "Coding Contest",
      description: "Compete with top coders",
      category: "Technical",
      image: "https://d1csarkz8obe9u.cloudfront.net/posterpreviews/coding-competition-design-template-87dba6fa6e8291b8fe5e29abc492288a_screen.jpg",
      date: "Feb 15 2026",
      time: "10:00 AM - 2:00 PM",
      venue: "Lab Block A",
      coordinator: "Dharmatej",
      inspector: "Prof. Meena, NIT Warangal"
    },
    {
      title: "Tech Workshop",
      description: "Learn new technologies",
      category: "Workshop",
      image: "https://img.freepik.com/free-photo/people-repairing-computer-chips_23-2150880942.jpg",
      date: "Feb 18 2026",
      time: "11:00 AM - 4:00 PM",
      venue: "Seminar Hall",
      coordinator: "B. Naveen Kumar",
      inspector: "Prof. Arjun, JNTU"
    },
    {
      title: "AI & ML Seminar",
      description: "Explore AI trends",
      category: "Technical",
      image: "https://aiml.sairam.edu.in/wp-content/uploads/sites/16/2023/11/5technical-1.1.jpg",
      date: "Feb 20 2026",
      time: "2:00 PM - 5:00 PM",
      venue: "Conference Hall",
      coordinator: "Ayush Pandya",
      inspector: "Prof. Kiran, IIIT Hyderabad"
    },
    {
      title: "Web Dev Bootcamp",
      description: "Full stack learning",
      category: "Workshop",
      image: "https://t4.ftcdn.net/jpg/08/61/69/45/360_F_861694590_3jqiyNFhOfL3LVpgmLQ7GmiNq6esOu6T.jpg",
      date: "Apr 22 2026",
      time: "10:00 AM - 5:00 PM",
      venue: "Computer Lab",
      coordinator: "Sanjay Karupothula",
      inspector: "Prof. Ravi, Osmania University"
    },
    {
      title: "Gaming Tournament",
      description: "Esports competition",
      category: "Sports",
      image: "https://gamespace.com/wp-content/uploads/2021/04/Tournaments.jpg",
      date: "Apr 25 2026",
      time: "12:00 PM - 6:00 PM",
      venue: "Indoor Stadium",
      coordinator: "Dharmatej",
      inspector: "Prof. Vivek, SR University"
    },
    {
      title: "Robotics Bootcamp",
      description: "Build and program robots",
      category: "Workshop",
      image: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRoHKIS4npZUnOIosoWywIylyuv3WZpG37QYg&s",
      date: "May 27 2026",
      time: "9:00 AM - 3:00 PM",
      venue: "Robotics Lab",
      coordinator: "B. Naveen Kumar",
      inspector: "Prof. Anil, IIT Madras"
    },
    {
      title: "Photography Contest",
      description: "Show your creativity",
      category: "Cultural",
      image: "https://images.unsplash.com/photo-1542038784456-1ea8e935640e",
      date: "Mar 21 2026",
      time: "10:00 AM - 1:00 PM",
      venue: "Campus Grounds",
      coordinator: "Ayush Pandya",
      inspector: "Prof. Sita, Fine Arts College"
    },
    {
      title: "Sports Meet",
      description: "Participate in athletic events",
      category: "Sports",
      image: "https://img.freepik.com/free-photo/sports-tools_53876-138077.jpg",
      date: "Apr 25 2026",
      time: "8:00 AM - 4:00 PM",
      venue: "Playground",
      coordinator: "Dharmatej",
      inspector: "Prof. Ramesh, Sports Authority"
    }
  ];

  useEffect(() => {
    const loadEvents = async () => {
      try {
        const response = await fetchEvents();
        const serverEvents = response.data.map((event) => ({
          ...event,
          id: event.id
        }));

        if (serverEvents.length > 0) {
          setEvents(serverEvents);
        } else {
          setEvents(defaultEvents);
        }
      } catch (error) {
        setEvents(defaultEvents);
      }
    };

    loadEvents();
  }, [eventRefreshKey]);

  // Find next upcoming event for hero section
  const getNextEvent = () => {
    const today = new Date();
    const upcomingEvents = events.filter(event => new Date(event.date) > today);
    if (upcomingEvents.length === 0) return null;
    return upcomingEvents.sort((a, b) => new Date(a.date) - new Date(b.date))[0];
  };

  const nextEvent = getNextEvent();

  // Countdown timer effect
  useEffect(() => {
    if (!nextEvent) return;

    const interval = setInterval(() => {
      const now = new Date().getTime();
      const eventTime = new Date(nextEvent.date).getTime();
      const distance = eventTime - now;

      if (distance > 0) {
        setCountdown({
          days: Math.floor(distance / (1000 * 60 * 60 * 24)),
          hours: Math.floor((distance % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60)),
          minutes: Math.floor((distance % (1000 * 60 * 60)) / (1000 * 60)),
          seconds: Math.floor((distance % (1000 * 60)) / 1000)
        });
      } else {
        setCountdown({ days: 0, hours: 0, minutes: 0, seconds: 0 });
      }
    }, 1000);

    return () => clearInterval(interval);
  }, [nextEvent]);

  const categories = ["All", "Technical", "Workshop", "Sports", "Cultural"];

  const today = new Date();
  const upcomingCount = events.filter((event) => {
    const eventDate = new Date(event.date);
    return eventDate > today;
  }).length;
  const completedCount = events.filter((event) => {
    const eventDate = new Date(event.date);
    return eventDate < today;
  }).length;
  const ongoingCount = events.filter((event) => {
    const eventDate = new Date(event.date);
    return eventDate.toDateString() === today.toDateString();
  }).length;

  let filteredEvents = events
    .filter((event) =>
      event.title.toLowerCase().includes(search.toLowerCase()) ||
      event.description.toLowerCase().includes(search.toLowerCase()) ||
      event.venue.toLowerCase().includes(search.toLowerCase())
    )
    .filter((event) =>
      category === "All" ? true : event.category === category
    )
    .filter((event) => {
      if (priceFilter === "free") return !event.isPaid;
      if (priceFilter === "paid") return event.isPaid;
      return true;
    })
    .filter((event) => {
      if (availabilityFilter === "available") {
        return !event.capacity || event.currentRegistrations < event.capacity;
      }
      if (availabilityFilter === "full") {
        return event.capacity && event.currentRegistrations >= event.capacity;
      }
      return true;
    });

  if (sort === "name") {
    filteredEvents.sort((a, b) => a.title.localeCompare(b.title));
  }

  if (sort === "date") {
    filteredEvents.sort((a, b) => new Date(a.date) - new Date(b.date));
  }

  if (sort === "status") {
    const getPriority = (event) => {
      const today = new Date();
      const eventDate = new Date(event.date);

      if (eventDate.toDateString() === today.toDateString()) return 1;
      if (eventDate > today) return 2;
      return 3;
    };

    filteredEvents.sort((a, b) => getPriority(a) - getPriority(b));
  }

  return (
    <div className="min-h-screen bg-gradient-to-br from-gray-950 via-purple-950 to-black p-6 md:p-10">
      <h1 className="text-4xl md:text-5xl font-bold text-white mb-6">
        UniSphere Events 🚀
      </h1>

      {/* Featured Hero Section */}
      {nextEvent && (
        <div className="mb-8 bg-gradient-to-r from-purple-600/20 to-cyan-600/20 border border-purple-500/30 rounded-3xl p-8 backdrop-blur-sm">
          <div className="flex flex-col md:flex-row items-center gap-6">
            <div className="flex-1">
              <h2 className="text-2xl md:text-3xl font-bold text-white mb-2">
                🔥 Next Big Event
              </h2>
              <h3 className="text-xl md:text-2xl font-semibold text-purple-300 mb-3">
                {nextEvent.title}
              </h3>
              <p className="text-gray-300 mb-4">{nextEvent.description}</p>
              <div className="flex flex-wrap gap-4 text-sm text-gray-400 mb-4">
                <span>📅 {nextEvent.date}</span>
                <span>📍 {nextEvent.venue}</span>
                <span>🏷️ {nextEvent.category}</span>
              </div>
              <button
                onClick={() => onRegister(nextEvent)}
                className="bg-gradient-to-r from-purple-500 to-cyan-500 hover:from-purple-600 hover:to-cyan-600 px-6 py-3 rounded-xl font-semibold transition-all duration-300 hover:scale-105"
              >
                Register Now 🚀
              </button>
            </div>
            <div className="flex-shrink-0">
              <div className="bg-black/50 rounded-2xl p-6 text-center">
                <h4 className="text-lg font-semibold text-white mb-4">⏰ Countdown</h4>
                <div className="grid grid-cols-4 gap-4">
                  <div className="bg-purple-600/20 rounded-lg p-3">
                    <div className="text-2xl font-bold text-white">{countdown.days}</div>
                    <div className="text-xs text-gray-400">Days</div>
                  </div>
                  <div className="bg-purple-600/20 rounded-lg p-3">
                    <div className="text-2xl font-bold text-white">{countdown.hours}</div>
                    <div className="text-xs text-gray-400">Hours</div>
                  </div>
                  <div className="bg-purple-600/20 rounded-lg p-3">
                    <div className="text-2xl font-bold text-white">{countdown.minutes}</div>
                    <div className="text-xs text-gray-400">Min</div>
                  </div>
                  <div className="bg-purple-600/20 rounded-lg p-3">
                    <div className="text-2xl font-bold text-white">{countdown.seconds}</div>
                    <div className="text-xs text-gray-400">Sec</div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      )}

      <div className="grid gap-4 md:grid-cols-3 mb-8">
        <div className="bg-white/10 border border-white/10 rounded-3xl p-5 backdrop-blur-sm">
          <p className="text-sm text-gray-400">Upcoming Events</p>
          <p className="text-3xl font-bold">{upcomingCount}</p>
        </div>
        <div className="bg-white/10 border border-white/10 rounded-3xl p-5 backdrop-blur-sm">
          <p className="text-sm text-gray-400">Ongoing Today</p>
          <p className="text-3xl font-bold">{ongoingCount}</p>
        </div>
        <div className="bg-white/10 border border-white/10 rounded-3xl p-5 backdrop-blur-sm">
          <p className="text-sm text-gray-400">Completed</p>
          <p className="text-3xl font-bold">{completedCount}</p>
        </div>
      </div>

      <div className="flex flex-wrap gap-4 mb-6">
        <input
          type="text"
          placeholder="Search events..."
          className="px-4 py-3 rounded-xl w-full max-w-md bg-gray-800 text-white placeholder-gray-400 outline-none focus:ring-2 focus:ring-purple-500"
          onChange={(e) => setSearch(e.target.value)}
        />

        <select
          onChange={(e) => setSort(e.target.value)}
          className="px-4 py-3 rounded-xl bg-gray-800 text-white"
        >
          <option value="default">Sort By</option>
          <option value="name">Name (A-Z)</option>
          <option value="date">Date</option>
          <option value="status">Status</option>
          <option value="popular">Most Popular</option>
          <option value="rating">Highest Rated</option>
        </select>

        <select
          onChange={(e) => setPriceFilter(e.target.value)}
          className="px-4 py-3 rounded-xl bg-gray-800 text-white"
        >
          <option value="all">All Prices</option>
          <option value="free">Free Only</option>
          <option value="paid">Paid Only</option>
        </select>

        <select
          onChange={(e) => setAvailabilityFilter(e.target.value)}
          className="px-4 py-3 rounded-xl bg-gray-800 text-white"
        >
          <option value="all">All Events</option>
          <option value="available">Available Spots</option>
          <option value="full">Full Events</option>
        </select>
      </div>

      <div className="flex flex-wrap gap-3 mb-8">
        {categories.map((cat) => (
          <button
            key={cat}
            onClick={() => setCategory(cat)}
            className={`px-6 py-3 rounded-xl font-medium transition-all duration-300 transform hover:scale-105 ${
              category === cat
                ? "bg-gradient-to-r from-purple-500 to-cyan-500 text-white shadow-lg"
                : "bg-white/10 text-gray-300 hover:bg-white/20 hover:text-white"
            }`}
          >
            {cat}
          </button>
        ))}
      </div>

      <div className="grid md:grid-cols-3 gap-8">
        {filteredEvents.map((event) => (
          <EventCard
            key={event.id}
            event={event}
            onRegister={onRegister}
          />
        ))}
      </div>
    </div>
  );
}
