// UniSphere/src/api.js
const USERS_KEY = "unisphere_users";
const EVENTS_KEY = "unisphere_events";
const REGISTRATIONS_KEY = "unisphere_registrations";

const defaultEvents = [
  {
    id: "event-1",
    title: "Hackathon 2026",
    description: "Build amazing projects",
    category: "Technical",
    image: "https://www.cpduk.co.uk/sites/default/files/news-imported/cpd-product-marketing-alliance-best-practices-hackathon.jpg",
    date: "2026-02-12",
    time: "9:00 AM - 6:00 PM",
    venue: "Main Auditorium",
    coordinator: "Sanjay Karupothula",
    inspector: "Prof. Rajesh, IIT Hyderabad",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-2",
    title: "Coding Contest",
    description: "Compete with top coders",
    category: "Technical",
    image: "https://d1csarkz8obe9u.cloudfront.net/posterpreviews/coding-competition-design-template-87dba6fa6e8291b8fe5e29abc492288a_screen.jpg",
    date: "2026-02-15",
    time: "10:00 AM - 2:00 PM",
    venue: "Lab Block A",
    coordinator: "Dharmatej",
    inspector: "Prof. Meena, NIT Warangal",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-3",
    title: "Tech Workshop",
    description: "Learn new technologies",
    category: "Workshop",
    image: "https://img.freepik.com/free-photo/people-repairing-computer-chips_23-2150880942.jpg",
    date: "2026-02-18",
    time: "11:00 AM - 4:00 PM",
    venue: "Seminar Hall",
    coordinator: "B. Naveen Kumar",
    inspector: "Prof. Arjun, JNTU",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-4",
    title: "AI & ML Seminar",
    description: "Explore AI trends",
    category: "Technical",
    image: "https://aiml.sairam.edu.in/wp-content/uploads/sites/16/2023/11/5technical-1.1.jpg",
    date: "2026-02-20",
    time: "2:00 PM - 5:00 PM",
    venue: "Conference Hall",
    coordinator: "Ayush Pandya",
    inspector: "Prof. Kiran, IIIT Hyderabad",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-5",
    title: "Web Dev Bootcamp",
    description: "Full stack learning",
    category: "Workshop",
    image: "https://t4.ftcdn.net/jpg/08/61/69/45/360_F_861694590_3jqiyNFhOfL3LVpgmLQ7GmiNq6esOu6T.jpg",
    date: "2026-04-22",
    time: "10:00 AM - 5:00 PM",
    venue: "Computer Lab",
    coordinator: "Sanjay Karupothula",
    inspector: "Prof. Ravi, Osmania University",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-6",
    title: "Gaming Tournament",
    description: "Esports competition",
    category: "Sports",
    image: "https://gamespace.com/wp-content/uploads/2021/04/Tournaments.jpg",
    date: "2026-04-25",
    time: "12:00 PM - 6:00 PM",
    venue: "Indoor Stadium",
    coordinator: "Dharmatej",
    inspector: "Prof. Vivek, SR University",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-7",
    title: "Robotics Bootcamp",
    description: "Build and program robots",
    category: "Workshop",
    image: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRoHKIS4npZUnOIosoWywIylyuv3WZpG37QYg&s",
    date: "2026-05-27",
    time: "9:00 AM - 3:00 PM",
    venue: "Robotics Lab",
    coordinator: "B. Naveen Kumar",
    inspector: "Prof. Anil, IIT Madras",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-8",
    title: "Photography Contest",
    description: "Show your creativity",
    category: "Cultural",
    image: "https://images.unsplash.com/photo-1542038784456-1ea8e935640e",
    date: "2026-03-21",
    time: "10:00 AM - 1:00 PM",
    venue: "Campus Grounds",
    coordinator: "Ayush Pandya",
    inspector: "Prof. Sita, Fine Arts College",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  },
  {
    id: "event-9",
    title: "Sports Meet",
    description: "Participate in athletic events",
    category: "Sports",
    image: "https://img.freepik.com/free-photo/sports-tools_53876-138077.jpg",
    date: "2026-04-25",
    time: "8:00 AM - 4:00 PM",
    venue: "Playground",
    coordinator: "Dharmatej",
    inspector: "Prof. Ramesh, Sports Authority",
    isPaid: false,
    price: 0,
    paymentMethods: ["stripe"],
    capacity: null,
    waitlistEnabled: false,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  }
];

const defaultUsers = [
  {
    id: "user-admin",
    name: "Admin",
    email: "admin@gmail.com",
    password: "admin123",
    role: "admin",
    phone: "",
    college: "",
    blocked: false,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  }
];

const parseJSON = (value, fallback) => {
  if (!value) return fallback;
  try {
    return JSON.parse(value);
  } catch {
    return fallback;
  }
};

const getData = (key, fallback) => {
  return parseJSON(localStorage.getItem(key), fallback);
};

const setData = (key, data) => {
  localStorage.setItem(key, JSON.stringify(data));
};

const getNextId = (prefix) => {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 6)}`;
};

const initStorage = () => {
  const users = getData(USERS_KEY, []);
  if (users.length === 0) {
    setData(USERS_KEY, defaultUsers);
  }

  const events = getData(EVENTS_KEY, []);
  if (events.length === 0) {
    setData(EVENTS_KEY, defaultEvents);
  }

  const registrations = getData(REGISTRATIONS_KEY, []);
  if (!Array.isArray(registrations)) {
    setData(REGISTRATIONS_KEY, []);
  }
};

const createResponse = (data) => ({ data });

export const registerUser = async (payload) => {
  initStorage();
  const users = getData(USERS_KEY, []);
  const normalizedEmail = payload.email.trim().toLowerCase();

  if (users.find((user) => user.email.toLowerCase() === normalizedEmail)) {
    throw new Error("User already exists");
  }

  const newUser = {
    id: getNextId("user"),
    name: payload.name,
    email: normalizedEmail,
    password: payload.password,
    role: "user",
    phone: payload.phone || "",
    college: payload.college || "",
    blocked: false,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  };

  users.push(newUser);
  setData(USERS_KEY, users);
  return createResponse(newUser);
};

export const loginUser = async (payload) => {
  initStorage();
  const users = getData(USERS_KEY, []);
  const normalizedEmail = payload.email.trim().toLowerCase();
  const user = users.find(
    (user) => user.email.toLowerCase() === normalizedEmail && user.password === payload.password
  );

  if (!user) {
    throw new Error("Invalid credentials");
  }

  if (user.blocked) {
    throw new Error("User is blocked");
  }

  return createResponse(user);
};

export const fetchCurrentUser = async (id) => {
  initStorage();
  const users = getData(USERS_KEY, []);
  const user = users.find((item) => item.id === id);
  if (!user) {
    throw new Error("User not found");
  }
  return createResponse(user);
};

export const updateProfile = async (id, payload) => {
  initStorage();
  const users = getData(USERS_KEY, []);
  const userIndex = users.findIndex((item) => item.id === id);
  if (userIndex === -1) {
    throw new Error("User not found");
  }
  users[userIndex] = {
    ...users[userIndex],
    name: payload.name ?? users[userIndex].name,
    phone: payload.phone ?? users[userIndex].phone,
    college: payload.college ?? users[userIndex].college,
    updatedAt: new Date().toISOString()
  };
  setData(USERS_KEY, users);
  return createResponse(users[userIndex]);
};

export const fetchEvents = async () => {
  initStorage();
  const events = getData(EVENTS_KEY, []);
  return createResponse(events);
};

export const addEvent = async (payload) => {
  initStorage();
  const events = getData(EVENTS_KEY, []);
  const newEvent = {
    id: getNextId("event"),
    title: payload.title,
    description: payload.description,
    category: payload.category,
    date: payload.date,
    time: payload.time || "TBD",
    venue: payload.venue,
    image: payload.image || "",
    coordinator: payload.coordinator || "",
    inspector: payload.inspector || "",
    isPaid: !!payload.isPaid,
    price: payload.isPaid ? (payload.price || 0) : 0,
    paymentMethods: payload.paymentMethods || ["stripe"],
    capacity: payload.capacity || null,
    waitlistEnabled: !!payload.waitlistEnabled,
    currentRegistrations: 0,
    status: "active",
    averageRating: 0,
    totalReviews: 0,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  };

  events.push(newEvent);
  setData(EVENTS_KEY, events);
  return createResponse(newEvent);
};

export const updateEvent = async (id, payload) => {
  initStorage();
  const events = getData(EVENTS_KEY, []);
  const eventIndex = events.findIndex((item) => item.id === id);
  if (eventIndex === -1) {
    throw new Error("Event not found");
  }

  events[eventIndex] = {
    ...events[eventIndex],
    title: payload.title ?? events[eventIndex].title,
    description: payload.description ?? events[eventIndex].description,
    category: payload.category ?? events[eventIndex].category,
    date: payload.date ?? events[eventIndex].date,
    time: payload.time ?? events[eventIndex].time,
    venue: payload.venue ?? events[eventIndex].venue,
    image: payload.image ?? events[eventIndex].image,
    coordinator: payload.coordinator ?? events[eventIndex].coordinator,
    inspector: payload.inspector ?? events[eventIndex].inspector,
    isPaid: payload.isPaid ?? events[eventIndex].isPaid,
    price: payload.isPaid ? (payload.price || 0) : 0,
    paymentMethods: payload.paymentMethods || events[eventIndex].paymentMethods,
    capacity: payload.capacity ?? events[eventIndex].capacity,
    waitlistEnabled: payload.waitlistEnabled ?? events[eventIndex].waitlistEnabled,
    updatedAt: new Date().toISOString()
  };

  setData(EVENTS_KEY, events);
  return createResponse(events[eventIndex]);
};

export const deleteEvent = async (id) => {
  initStorage();
  let events = getData(EVENTS_KEY, []);
  let registrations = getData(REGISTRATIONS_KEY, []);
  events = events.filter((item) => item.id !== id);
  registrations = registrations.filter((registration) => registration.eventId !== id);
  setData(EVENTS_KEY, events);
  setData(REGISTRATIONS_KEY, registrations);
  return createResponse({ message: "Event deleted successfully" });
};

export const registerEvent = async (payload) => {
  initStorage();
  const users = getData(USERS_KEY, []);
  const events = getData(EVENTS_KEY, []);
  const registrations = getData(REGISTRATIONS_KEY, []);

  const user = users.find((item) => item.id === payload.userId);
  if (!user) {
    throw new Error("User not found");
  }

  const event = events.find((item) => item.id === payload.eventId);
  if (!event) {
    throw new Error("Event not found");
  }

  if (event.status !== "active") {
    throw new Error("Event is not available for registration");
  }

  if (registrations.find((item) => item.userId === payload.userId && item.eventId === payload.eventId)) {
    throw new Error("Already registered for this event");
  }

  let registrationStatus = "confirmed";
  let waitlistPosition = null;

  if (event.capacity && event.currentRegistrations >= event.capacity) {
    if (!event.waitlistEnabled) {
      throw new Error("Event is full");
    }
    registrationStatus = "waitlisted";
    const existingWaitlist = registrations.filter(
      (item) => item.eventId === payload.eventId && item.status === "waitlisted"
    ).length;
    waitlistPosition = existingWaitlist + 1;
  }

  const newRegistration = {
    id: getNextId("registration"),
    userId: payload.userId,
    eventId: payload.eventId,
    eventTitle: payload.eventTitle,
    status: registrationStatus,
    waitlistPosition,
    registeredAt: new Date().toISOString(),
    rating: null,
    review: "",
    reviewedAt: null
  };

  registrations.push(newRegistration);
  if (registrationStatus === "confirmed") {
    event.currentRegistrations = (event.currentRegistrations || 0) + 1;
    event.updatedAt = new Date().toISOString();
  }

  setData(REGISTRATIONS_KEY, registrations);
  setData(EVENTS_KEY, events);

  return createResponse({ ...newRegistration, message: registrationStatus === "waitlisted" ? `Added to waitlist (Position ${waitlistPosition})` : "Successfully registered!" });
};

export const fetchRegistrations = async (userId) => {
  initStorage();
  const registrations = getData(REGISTRATIONS_KEY, []);
  const events = getData(EVENTS_KEY, []);

  const userRegistrations = registrations
    .filter((item) => item.userId === userId)
    .map((registration) => ({
      ...registration,
      eventId: events.find((event) => event.id === registration.eventId) || null
    }));

  return createResponse(userRegistrations);
};

export const fetchAllRegistrations = async () => {
  initStorage();
  const registrations = getData(REGISTRATIONS_KEY, []);
  const events = getData(EVENTS_KEY, []);

  const allRegistrations = registrations.map((registration) => ({
    ...registration,
    eventId: events.find((event) => event.id === registration.eventId) || null
  }));

  return createResponse(allRegistrations);
};

export const fetchUsers = async () => {
  initStorage();
  const users = getData(USERS_KEY, []);
  return createResponse(users.map((user) => {
    const { password: _, ...userWithoutPassword } = user;
    return userWithoutPassword;
  }));
};

export const blockUser = async (id) => {
  initStorage();
  const users = getData(USERS_KEY, []);
  const userIndex = users.findIndex((item) => item.id === id);
  if (userIndex === -1) {
    throw new Error("User not found");
  }
  users[userIndex].blocked = true;
  users[userIndex].updatedAt = new Date().toISOString();
  setData(USERS_KEY, users);
  return createResponse({ message: "User blocked" });
};

export const unblockUser = async (id) => {
  initStorage();
  const users = getData(USERS_KEY, []);
  const userIndex = users.findIndex((item) => item.id === id);
  if (userIndex === -1) {
    throw new Error("User not found");
  }
  users[userIndex].blocked = false;
  users[userIndex].updatedAt = new Date().toISOString();
  setData(USERS_KEY, users);
  return createResponse({ message: "User unblocked" });
};

export const rateEvent = async (registrationId, rating, review) => {
  initStorage();
  const registrations = getData(REGISTRATIONS_KEY, []);
  const events = getData(EVENTS_KEY, []);
  const registrationIndex = registrations.findIndex((item) => item.id === registrationId);
  if (registrationIndex === -1) {
    throw new Error("Registration not found");
  }

  registrations[registrationIndex] = {
    ...registrations[registrationIndex],
    rating,
    review,
    reviewedAt: new Date().toISOString()
  };

  const event = events.find((item) => item.id === registrations[registrationIndex].eventId);
  if (event) {
    const eventRatings = registrations.filter(
      (item) => item.eventId === event.id && typeof item.rating === "number"
    );
    const totalRating = eventRatings.reduce((sum, item) => sum + item.rating, 0);
    event.averageRating = eventRatings.length > 0 ? Math.round((totalRating / eventRatings.length) * 10) / 10 : 0;
    event.totalReviews = eventRatings.length;
    event.updatedAt = new Date().toISOString();
  }

  setData(REGISTRATIONS_KEY, registrations);
  setData(EVENTS_KEY, events);

  return createResponse(registrations[registrationIndex]);
};

export const createPaymentSession = async () => {
  throw new Error("Payments are not supported in local mode.");
};

export const processPayment = async () => {
  throw new Error("Payments are not supported in local mode.");
};

export const getPaymentDetails = async () => {
  throw new Error("Payments are not supported in local mode.");
};

export const getUserPayments = async () => {
  throw new Error("Payments are not supported in local mode.");
};

export const refundPayment = async () => {
  throw new Error("Payments are not supported in local mode.");
};

export const getPaymentStats = async () => {
  throw new Error("Payments are not supported in local mode.");
};
